### ISSUE TITLE
refactor: Extract abstract iteration patterns from numerical/coupling/ to iterative/

### ISSUE THREAD

**Issue by derrring** [derrring/MFG_PDE#630]
## Summary

Extract abstract iteration patterns (Picard, Fictitious Play, Newton) from `alg/numerical/coupling/` to `alg/iterative/` for reuse across paradigms (numerical, neural, RL).

## Motivation

Current `numerical/coupling/` solvers contain both:
1. **Abstract iteration logic** (convergence checking, damping, learning rate schedules)
2. **Numerical-specific implementation** (grid spacing, L2 norms, HJB/FP interfaces)

The abstract patterns could be reused by:
- `neural/` - PINN with Picard coupling
- `reinforcement/` - Policy Iteration (structurally identical to Picard)
- `optimization/` - Variational methods with alternating minimization

## Proposed Structure

```
alg/
├── iterative/                    # Abstract layer - iteration patterns
│   ├── base.py                   # BaseIterativeSolver (exists)
│   ├── patterns/
│   │   ├── picard.py            # PicardPattern - abstract Picard iteration
│   │   ├── fictitious_play.py   # FictitiousPlayPattern - decaying LR
│   │   └── newton.py            # NewtonPattern - quadratic convergence
│   └── convergence.py           # Generic convergence criteria
│
├── numerical/coupling/           # Numerical implementations
│   ├── fixed_point_iterator.py  # Inherits PicardPattern
│   ├── fictitious_play.py       # Inherits FictitiousPlayPattern
│   └── newton_mfg_solver.py     # Inherits NewtonPattern
│
├── neural/coupling/              # Neural implementations (future)
│   └── pinn_picard.py           # PINN + Picard
│
└── reinforcement/coupling/       # RL implementations (future)
    └── policy_iteration.py      # Policy iteration ≈ Picard
```

### 1. PicardPattern
```python
class PicardPattern(Protocol):
    """Abstract Picard (fixed-point) iteration pattern."""
    
    def compute_forward_step(self, state: T) -> T:
        """Compute L(state) - the forward operator."""
        ...
    
    def apply_damping(self, old: T, new: T, alpha: float) -> T:
        """Apply damping: (1-α)·old + α·new"""
        ...
    
    def check_convergence(self, old: T, new: T) -> tuple[bool, str]:
        """Check if iteration has converged."""
        ...
```

### 2. FictitiousPlayPattern
```python
class FictitiousPlayPattern(PicardPattern):
    """Fictitious Play with decaying learning rate."""
    
    def get_learning_rate(self, iteration: int) -> float:
        """Compute α(k) = 1/(k+1) or custom schedule."""
        ...
```

### 3. Convergence Utilities
- `calculate_relative_change(old, new, metric)` - generic
- `calculate_l2_change(old, new, dx, dt)` - numerical-specific

## Implementation Plan

1. [ ] Create `alg/iterative/patterns/` directory
2. [ ] Extract `PicardPattern` Protocol from `FixedPointIterator`
3. [ ] Extract `FictitiousPlayPattern` from `FictitiousPlayIterator`
4. [ ] Extract learning rate schedules to `iterative/schedules.py`
5. [ ] Refactor `numerical/coupling/` to inherit from patterns
6. [ ] Add type hints ensuring paradigm-specific implementations

## Benefits

- **Code reuse**: One Picard implementation, multiple paradigms
- **Consistency**: Same convergence API across all solvers
- **Extensibility**: Easy to add new paradigms (neural, RL)
- **Testing**: Test abstract patterns once, implementations separately

## References

- Issue #628: Consolidated `solvers/` into `alg/iterative/`
- RL Policy Iteration is structurally identical to MFG Picard iteration
- Cardaliaguet & Hadikhanloo (2017) - Fictitious Play for MFG
